test/utils.cpp

```diff
@@ -338,6 +338,80 @@ TensorView* matmul(
   return matmulTuringOrLater(a, b, layout);
 }
 
+TensorView* matmulTuringOrLater_v2(
+    TensorView* a,
+    TensorView* b,
+    MmaLayout layout) {
+  NVF_CHECK(a->nDims() == b->nDims());
+  NVF_CHECK(a->nDims() == 2 || a->nDims() == 3);
+  TensorView *tv3 = nullptr, *tv2 = nullptr, *tv0t = nullptr, *tv1t = nullptr,
+             *tv0b = nullptr, *tv1b = nullptr;
+  if (a->nDims() == 3) { // bmm
+    switch (layout) {
+        // Canonicalize all inputs to [B, M, K] and [B, N, K]
+      case MmaLayout::TT:
+        tv0t = transpose(a, 0, 1);
+        tv1t = transpose(b, 1, 2);
+        break;
+      case MmaLayout::TN:
+        tv0t = transpose(a, 0, 1);
+        tv1t = transpose(b, 0, 1);
+        break;
+      case MmaLayout::NT:
+        tv0t = transpose(a, 1, 2);
+        tv1t = transpose(b, 1, 2);
+        break;
+      case MmaLayout::NN:
+        tv0t = transpose(a, 1, 2);
+        tv1t = transpose(b, 0, 1);
+        break;
+      default:
+        NVF_CHECK(false, "unsupported data layout.");
+    }
+  } else {
+    switch (layout) {
+        // Canonicalize all inputs to [M, K] and [N, K]
+      case MmaLayout::TT:
+        tv0t = a;
+        tv1t = transpose(b, 0, 1);
+        break;
+      case MmaLayout::TN:
+        tv0t = a;
+        tv1t = b;
+        break;
+      case MmaLayout::NT:
+        tv0t = transpose(a, 0, 1);
+        tv1t = transpose(b, 0, 1);
+        break;
+      case MmaLayout::NN:
+        tv0t = transpose(a, 0, 1);
+        tv1t = b;
+        break;
+      default:
+        NVF_CHECK(false, "unsupported data layout.");
+    }
+  }
+  std::vector<bool> bcast_dims(a->nDims() + 1, false);
+  bcast_dims.at(bcast_dims.size() - 2) = true;
+  tv0b = broadcast(tv0t, bcast_dims);
+  bcast_dims.at(bcast_dims.size() - 2) = false;
+  bcast_dims.at(bcast_dims.size() - 3) = true;
+  tv1b = broadcast(tv1t, bcast_dims);
+  tv2 = mul(tv0b, tv1b);
+  tv3 = sum(tv2, {-1});
+  // tv2 = fusedMultiplySum(tv0b, tv1b, {-1});
```

Comment:
Since this is the extent of this change, could it be implemented as a flag in `matmul` instead of a separate function in order to reduce code duplication?

Reply:
More than happy to. I wasn't sure if we wanted to mix implementations.